controller/volume_controller.go

```diff
@@ -250,11 +250,17 @@ func (vc *VolumeController) syncVolume(key string) (err error) {
 
 	log := getLoggerForVolume(vc.logger, volume)
 
+	// we want to always check whether we should be the owner of this volume
+	// otherwise we might end up doing an iteration while we are in the process of
+	// an ownership transfer (v.Spec.NodeID != v.Status.CurrentOwnerID)
+	// this will lead to 2 controllers fighting about the same object and a couple of conflicts on the volume updates.
+	// The real problem could be the updates that are run on the dependent components (engine, replicas, etc)
+	// by the no longer valid owner of this volume
```

Comment:
More or less yes, `isResponsible` should always be run each loop iteration for all resources.
And `Status.OwnerID != controllerID` should then only be responsible for updating the ownership status.

This allows for voluntary ownership transfers.
Right now if a ownership transfer from node1 to node2 needs to happen because of v.Spec.NodeID changes.
Than the previous owner will process that loop iteration while the future owner will try to claim ownership.
This leads to a race between the two.

Since updates happen at the end of the loop iteration, the conflict will only be caught then.
But in the case where a controller updates dependent components, this will lead to the previous owner potentially changing the state of the dependent component, since it is racing with the future owner.

Does that explanation make sense?

Reply:
Yes. That's more clear.